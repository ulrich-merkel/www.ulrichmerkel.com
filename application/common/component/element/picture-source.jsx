/* eslint-disable immutable/no-mutation */
/**
 * Rendering a source html tag for the use in a picture element.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @requires react
 * @requires prop-types
 * @requires classnames
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 */
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {Object} props - The current component props
 * @param {string} props.ext - The image extension (jpg, png)
 * @param {string} props.name - The image name
 * @param {string} props.path - The image path (folder)
 * @param {string|number} props.width - The image width (for dynamically creating names)
 * @param {string|number} props.height - The image height (for dynamically creating names)
 * @param {string|number} props.minWidth - The mediaquery min-width value
 * @param {string} [props.className] - The component css class names, will be merged into component default classNames
 * @returns {React.Element} React component markup
 */
function ElementPictureSource(props) {
    const {
        className,
        ext,
        height,
        minWidth,
        name,
        path,
        width
    } = props;

    const composedClassName = classnames(
        'c-picture__source',
        className
    );

    const srcSet = `${path}${name}@${width}x${height}.${ext}`;
    const media = `(min-width: ${minWidth}px)`;

    return (
        <source className={composedClassName} {...{ srcSet, media }} />
    );
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {Object}
 */
ElementPictureSource.propTypes = {
    ext: PropTypes.oneOf([
        'jpg',
        'png',
        ''
    ]).isRequired,
    name: PropTypes.string.isRequired,
    path: PropTypes.string.isRequired,
    width: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number
    ]).isRequired,
    height: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number
    ]).isRequired,
    minWidth: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number
    ]).isRequired,
    className: PropTypes.string // eslint-disable-line react/require-default-props
};

export default ElementPictureSource;
