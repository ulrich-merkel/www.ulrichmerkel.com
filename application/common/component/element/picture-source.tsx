/**
 * Rendering a source html tag for the use in a picture element.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 *
 * @see {@link https://css-tricks.com/using-webp-images/}
 */
import { default as React, Fragment, FunctionComponent } from 'react';
import classnames from 'classnames';

type Props = {
    ext: 'jpg' | 'png' | '';
    name: string;
    path: string;
    width: number;
    height: number;
    minWidth: number;
    className?: string;
};

/**
 * Small helper to get correct type attribute.
 *
 * @private
 * @param {string} ext - The file extension
 * @returns {string} The source element's type description
 */
function getType(ext: string) {
    switch (ext) {
        case 'jpg':
            return 'image/jpeg';
        default:
            return `image/${ext}`;
    }
}

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} props - The current component props
 * @param {string} props.ext - The image extension (jpg, png)
 * @param {string} props.name - The image name
 * @param {string} props.path - The image path (folder)
 * @param {string|number} props.width - The image width (for dynamically creating names)
 * @param {string|number} props.height - The image height (for dynamically creating names)
 * @param {string|number} props.minWidth - The mediaquery min-width value
 * @param {string} [props.className] - The component css class names, will be merged into component default classNames
 * @returns {ReactElement} React component markup
 */
export const PictureSource: FunctionComponent<Props> = (props) => {
    const { className, ext, height, minWidth, name, path, width } = props;

    const composedClassName = classnames('c-picture__source', className);

    const srcSet = `${path}${name}@${width}x${height}`;
    const media = `(min-width: ${minWidth}px)`;

    return (
        <Fragment>
            <source
                className={composedClassName}
                key={`${srcSet}.webp`}
                media={media}
                srcSet={`${srcSet}.webp`}
                type={getType('webp')}
            />
            <source
                className={composedClassName}
                key={`${srcSet}.${ext}`}
                media={media}
                srcSet={`${srcSet}.${ext}`}
                type={getType(ext)}
            />
        </Fragment>
    );
}
