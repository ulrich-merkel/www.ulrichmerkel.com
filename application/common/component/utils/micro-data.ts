/**
 * Es6 module for creating mirco data attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type ItemTypeAttributes = {
    itemScope: boolean;
    itemType: string;
};

/**
 * Get component properties for itemProp and itemType.
 *
 * @param {boolean} itemType - The itemType property if available
 * @returns {object|null} The component props for spreading
 */
export function getItemTypeAttributes(
    itemType?: string
): ItemTypeAttributes | null {
    return isValidString(itemType)
        ? {
              itemScope: true,
              itemType
          }
        : null;
}
