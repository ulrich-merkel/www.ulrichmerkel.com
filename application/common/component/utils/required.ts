/**
 * Es6 module for creating form input required attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isBoolean } from 'lodash';

type RequiredAttributes = {
    'aria-required': boolean;
    required: boolean;
};

/**
 * Get component properties for required form input elements.
 *
 * @param {boolean} required - Whether the input element is required or not
 * @returns {object|null} The component props for spreading
 */
export function getRequiredAttributes(
    required?: boolean
): RequiredAttributes | null {
    return isBoolean(required) && required
        ? {
              'aria-required': true,
              required: true
          }
        : null;
}
