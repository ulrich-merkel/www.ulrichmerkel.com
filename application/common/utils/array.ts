/**
 * Es6 module for handling array related utilities.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */

/**
 * Check if given "input" is a valid array with at least one entry.
 *
 * @param {Array<*>} array - The array to be checked
 * @returns {boolean} Whether this is an array with some entries
 */
export function isValidArray(array?: any[]): boolean {
    return Array.isArray(array) && array.length > 0;
}
