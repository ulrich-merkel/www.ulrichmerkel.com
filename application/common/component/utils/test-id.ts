/**
 * Es6 module for creating test id related attributes
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type TestIdAttributes = {
    'data-testid': string;
};

/**
 * Get component properties for testing.
 *
 * @param {string} testId - The test id to be added
 * @returns {object|null} The component props for spreading
 */
export function getTestIdAttributes(testId?: string): TestIdAttributes | null {
    return isValidString(testId)
        ? {
              'data-testid': testId
          }
        : null;
}
