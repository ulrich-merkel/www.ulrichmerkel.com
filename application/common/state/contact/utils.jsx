/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 *
 * @TODO Should be refactored and moved to actions
 *
 * @requires common/utils/validation
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';

/**
 * Validation helper to check all inputs.
 *
 * @function
 * @param {object} [state={}] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
function validate(state = {}) {
    return {
        name: required(state.name),
        email: required(state.email) && email(state.email),
        subject: required(state.subject),
        message: required(state.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @function
 * @param {object} state - The current redux contact state
 * @returns {boolean} The validation status
 */
function isValid(state) {
    const validated = validate(state);

    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

function canSendForm(state) {
    const isStateValid = isValid(state);

    return isStateValid
        && state.pristine
        && !state.pending
        && isBrowser(state);
}

export {
    canSendForm,
    isValid,
    validate
};
