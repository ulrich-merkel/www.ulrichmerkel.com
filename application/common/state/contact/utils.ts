/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';
import { ContactFormType, ValidateType } from './types';

/**
 * Validation helper to check all inputs.
 *
 * @param {object} [contactForm] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
export function validate(contactForm: ContactFormType): ValidateType {
    return {
        name: required(contactForm?.name),
        email: required(contactForm?.email) && email(contactForm?.email),
        subject: required(contactForm?.subject),
        message: required(contactForm?.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} The validation status
 */
export function isValid(contactForm: ContactFormType): boolean {
    const validated = validate(contactForm);
console.log({validated, contactForm})
    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

/**
 * Check if the form can be send.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} Whether the submit button is enabled or not
 */
export function canSendForm(contactForm: ContactFormType): boolean {
    const isStateValid = isValid(contactForm);

    return (
        isStateValid &&
        contactForm.pristine &&
        !contactForm.pending &&
        isBrowser(contactForm)
    );
}
