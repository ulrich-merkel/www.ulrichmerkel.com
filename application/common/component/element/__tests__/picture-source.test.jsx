import * as React from 'react';
import { render } from '../../../../__tests__/utils/test-utils';
import { PictureSource } from '../picture-source';

describe('PictureSource', function fnDescribe() {
    it('should render correctly', function fnIt() {
        const { asFragment } = render(
            <PictureSource
                path="path"
                name="name"
                ext="jpg"
                width="200"
                height="100"
                minWidth="100"
                className="test"
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
