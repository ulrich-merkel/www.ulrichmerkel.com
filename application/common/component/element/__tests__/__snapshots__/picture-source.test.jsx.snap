// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PictureSource should render correctly 1`] = `
<DocumentFragment>
  <source
    class="c-picture__source test"
    media="(min-width: 100px)"
    srcset="pathname@200x100.webp"
    type="image/webp"
  />
  <source
    class="c-picture__source test"
    media="(min-width: 100px)"
    srcset="pathname@200x100.jpg"
    type="image/jpeg"
  />
</DocumentFragment>
`;
